### PR TITLE
Topmenu

### PR DESCRIPTION
**Pull request by psim42**
Fixed a lot of lint/jest bugs, added i18n everywhere and home redirect on title.

### CODE REVIEW

File: client/components/Layout/LanguageButton.style.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {
+  createStyles,
+  makeStyles,
+
```
Comment (jde-maga) [inline client/components/Layout/LanguageButton.style.tsx]: remove \n

File: client/components/Layout/LanguageButton.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import Button from '@material-ui/core/Button';
+import Menu from '@material-ui/core/Menu';
+import MenuItem from '@material-ui/core/MenuItem';
+import { useIntl } from 'react-intl';
+import useStyles from './LanguageButton.style';
```
Comment (jde-maga) [inline client/components/Layout/LanguageButton.tsx]: add \n between modules own files

File: client/components/Layout/LanguageButton.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import { Locale } from '../../models/models';
+
+const options = [
+  'fr',
+  'en',
+];
+
+export default function SimpleMenu({ locale, setLocale }: Locale) {
```
Comment (jde-maga) [inline client/components/Layout/LanguageButton.tsx]: use fat arrow function

File: client/components/Layout/LanguageButton.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const classes = useStyles({});
+  const { formatMessage: _t } = useIntl();
+  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+  const [selectedIndex, setSelectedIndex] = React.useState(1);
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
```
Comment (jde-maga) [inline client/components/Layout/LanguageButton.tsx]: add \n between variables and functions

File: client/components/Layout/TopMenu.tsx [hunk 5]
```diff
@@ -1,38 +1,117 @@
...
+//   const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+
+//   const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+//     setAnchorEl(event.currentTarget);
+//   };
+
+//   const handleClose = () => {
+//     setAnchorEl(null);
+//   };
+// }
```
Comment (jde-maga) [inline client/components/Layout/TopMenu.tsx]: remove if unnecessary

File: client/components/Layout/TopMenu.tsx [hunk 6]
```diff
@@ -1,38 +1,117 @@
...
+import MenuItem from '@material-ui/core/MenuItem';
+import Menu from '@material-ui/core/Menu';
 import MenuIcon from '@material-ui/icons/Menu';
 import SearchIcon from '@material-ui/icons/Search';
+import AccountCircle from '@material-ui/icons/AccountCircle';
+import MailIcon from '@material-ui/icons/Mail';
+import NotificationsIcon from '@material-ui/icons/Notifications';
+import MoreIcon from '@material-ui/icons/MoreVert';
 import { useIntl } from 'react-intl';
+import history from '../../helpers/history';
```
Comment (jde-maga) [inline client/components/Layout/TopMenu.tsx]: \n between modules and own files

File: client/components/Layout/TopMenu.tsx [hunk 7]
```diff
@@ -1,38 +1,117 @@
...
+//     setAnchorEl(null);
+//   };
+// }
+
+export default function PrimarySearchAppBar({ locale, setLocale }: Locale): void {
   const { formatMessage: _t } = useIntl();
   const classes = useStyles({});
+  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+  const [mobileMoreAnchorEl, setMobileMoreAnchorEl] = React.useState<null | HTMLElement>(null);
+
```
Comment (jde-maga) [inline client/components/Layout/TopMenu.tsx]: remove \n in between variables

File: client/components/Layout/TopMenu.tsx [hunk 8]
```diff
@@ -1,38 +1,117 @@
...
+  const handleMenuClose = (): void => {
+    setAnchorEl(null);
+    handleMobileMenuClose();
+  };
+
+  const handleMobileMenuOpen = (event: React.MouseEvent<HTMLElement>): void => {
+    setMobileMoreAnchorEl(event.currentTarget);
+  };
+
+  const menuId = 'primary-search-account-menu';
```
Comment (jde-maga) [inline client/components/Layout/TopMenu.tsx]: move variable on top

File: false/_logs/2019-11-15T18_34_49_346Z-debug.log [hunk 9]
```diff
@@ -0,0 +1,33 @@
+0 info it worked if it ends with ok
```
Comment (jde-maga) [inline false/_logs/2019-11-15T18_34_49_346Z-debug.log]: remove log file

File: false/_logs/2019-11-15T18_35_09_737Z-debug.log [hunk 10]
```diff
@@ -0,0 +1,33 @@
+0 info it worked if it ends with ok
```
Comment (jde-maga) [inline false/_logs/2019-11-15T18_35_09_737Z-debug.log]: remove log file

File: false/_logs/2019-11-15T18_35_50_714Z-debug.log [hunk 11]
```diff
@@ -0,0 +1,40 @@
+0 info it worked if it ends with ok
```
Comment (jde-maga) [inline false/_logs/2019-11-15T18_35_50_714Z-debug.log]: remove log file

File: false/_logs/2019-11-15T18_35_58_309Z-debug.log [hunk 12]
```diff
@@ -0,0 +1,40 @@
+0 info it worked if it ends with ok
```
Comment (jde-maga) [inline false/_logs/2019-11-15T18_35_58_309Z-debug.log]: remove log file

File: client/components/Layout/LanguageButton.tsx (hunk 2 above)
Comment (psim42) [reply client/components/Layout/LanguageButton.tsx]: I don't understand this one ?

File: client/components/Layout/TopMenu.tsx [hunk 13]
```diff
@@ -1,38 +1,117 @@
...
-            className={classes.menuButton}
-            color="inherit"
-            aria-label="open drawer"
-          >
-            <MenuIcon />
-          </IconButton>
-          <Typography className={classes.title} variant="h6" noWrap>
+          <Typography onClick={(): void => redirectTitle()} className={classes.title} variant="h6" noWrap>
             {_t({ id: 'title' })}
           </Typography>
```
Comment (FlorianMuller) [inline client/components/Layout/TopMenu.tsx]: add css `cursor: pointer;` to the title (without user can't know he can click on it)

File: client/components/Layout/index.tsx [hunk 14]
```diff
@@ -11,12 +11,14 @@ import TopMenu from './TopMenu';
 
 interface Props {
   children: ReactElement;
+  locale: string;
+  setLocale: (locale: string) => void;
 }
 
-const Layout = ({ children }: Props): ReactElement => (
+const Layout = ({ children, locale, setLocale }: Props): ReactElement => (
```
Comment (FlorianMuller) [inline client/components/Layout/index.tsx]: Maybe change this file name to `Layout.tsx`, as it implement the `Layout` component. I find the name `index.tsx` quite confusing

File: client/components/Layout/index.tsx [hunk 15]
```diff
@@ -11,12 +11,14 @@ import TopMenu from './TopMenu';
...
   <div>
 
-    <TopMenu />
+    <TopMenu locale={locale} setLocale={setLocale} />
```
Comment (FlorianMuller) [inline client/components/Layout/index.tsx]: Add `{children}` here, without it, things between `<Layout> </Layout>` are not rendered :(

File: client/components/Layout/LanguageButton.tsx (hunk 2 above)
Comment (psim42) [reply client/components/Layout/LanguageButton.tsx]: Added \n after line 6 but I'm not sure that's what you mean

File: client/components/Layout/TopMenu.styles.tsx [hunk 16]
```diff
@@ -7,14 +7,13 @@ import {
 } from '@material-ui/core/styles';
```
[the remark this answers is not in the document]
Comment (psim42) [reply client/components/Layout/TopMenu.styles.tsx]: ???

File: client/components/Layout/index.tsx (hunk 14 above)
Comment (psim42) [reply client/components/Layout/index.tsx]: I didn't make this file, but I agree that the filename is wrong

File: client/components/Layout/LanguageButton.tsx [hunk 17]
```diff
@@ -0,0 +1,56 @@
+import React, { ReactElement } from "react";
+import Button from "@material-ui/core/Button";
+import Menu from "@material-ui/core/Menu";
+import MenuItem from "@material-ui/core/MenuItem";
+
```
Comment (jde-maga) [inline client/components/Layout/LanguageButton.tsx]: Breaking liiiiine

File: client/components/Layout/TopMenu.tsx [hunk 18]
```diff
@@ -4,28 +4,102 @@ import Toolbar from "@material-ui/core/Toolbar";
...
 import { useIntl } from "react-intl";
+
+import history from "../../helpers/history";
+
 import useStyles from "./TopMenu.styles";
+import LanguageButton from "./LanguageButton";
+import { Locale } from "../../models/models";
 
-export default function SearchAppBar() {
+export default function PrimarySearchAppBar({
```
Comment (FlorianMuller) [inline client/components/Layout/TopMenu.tsx]: Use arrow function notation, like in the rest of the code (<3 `() => {}` <3)

AND rename component `PrimarySearchAppBar ` to `TopMenu` (as it's name like that in the rest of the code and it's file is named also like that)

File: client/components/Layout/TopMenu.tsx [hunk 19]
```diff
@@ -4,28 +4,102 @@ import Toolbar from "@material-ui/core/Toolbar";
...
+        </IconButton>
+        <p>Profile</p>
+      </MenuItem>
+    </Menu>
+  );
 
   return (
-    <div className={classes.root}>
-      <AppBar position="static" color="primary">
+    <div className={classes.grow} style={{ margin: 0 }}>
```
Comment (FlorianMuller) [inline client/components/Layout/TopMenu.tsx]: I thinks it's better to put css in a class in the `style.tsx` than directly in the `style` parameter.
But maybe there's a justification i don't see here ? If not, move `style={{ margin: 0 }}` to a class

File: client/models/models.ts [hunk 20]
```diff
@@ -1,3 +1,7 @@
+export interface Locale {
+  locale: string;
+  setLocale: (locale: string) => void;
+}
```
Comment (FlorianMuller) [inline client/models/models.ts]: I'm not sure here, but if the interface `Locale` is not use in other part of the code, maybe it should be in the `component/Layout` directory ?
@jde-maga @psim42 @Jcharloi

File: client/components/Layout/index.tsx [hunk 21]
```diff
@@ -1,14 +1,20 @@
 import React, { ReactElement } from "react";
```
[the remark this answers is not in the document]
Comment (jde-maga) [reply client/components/Layout/index.tsx]: We should keep the index name to differentiate the Component used inside the file (names like Layout) from the actual main component (index)

File: client/models/models.ts (hunk 20 above)
Comment (jde-maga) [reply client/models/models.ts]: Keeping external interfaces (not Props, State) in a different folder might be better to keep imports consistent

**Comment by jde-maga** [comment]
The choice on top is good but we need to focus on better features for now, add a note in trello to not forget it

We should not use Context but rather a hook, but then again we need to focus on better features. Add it too

File: client/components/Layout/TopMenu.tsx [hunk 22]
```diff
@@ -4,28 +4,99 @@ import Toolbar from "@material-ui/core/Toolbar";
 import IconButton from "@material-ui/core/IconButton";
 import Typography from "@material-ui/core/Typography";
 import InputBase from "@material-ui/core/InputBase";
-import MenuIcon from "@material-ui/icons/Menu";
+import MenuItem from "@material-ui/core/MenuItem";
+import Menu from "@material-ui/core/Menu";
 import SearchIcon from "@material-ui/icons/Search";
+import AccountCircle from "@material-ui/icons/AccountCircle";
 import { useIntl } from "react-intl";
```
Comment (FlorianMuller) [inline client/components/Layout/TopMenu.tsx]: don't we put empty line between library imports and hooks imports ? (as you've done in `LanguageButton.tsx`)

File: client/components/Layout/TopMenu.tsx [hunk 23]
```diff
@@ -4,28 +4,99 @@ import Toolbar from "@material-ui/core/Toolbar";
...
+
+import history from "../../helpers/history";
+
 import useStyles from "./TopMenu.styles";
```
Comment (FlorianMuller) [inline client/components/Layout/TopMenu.tsx]: I think `useStyle` hooks whould be imports with next to `useIntl`

File: client/components/Layout/LanguageButton.tsx [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+          className={classes.languageMenu}
+          onClick={(): null => handleClose("fr")}
+        >
+          {_t({ id: "language.fr" })}
+        </MenuItem>
+        <MenuItem
+          className={classes.languageMenu}
+          onClick={(): null => handleClose("en")}
+        >
+          {_t({ id: "language.en" })}
```
Comment (FlorianMuller) [inline client/components/Layout/LanguageButton.tsx]: Eslint advise to put emoji into span: [more info here](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/accessible-emoji.md)

Maybe we could do something like:
```html
<span role="img" aria-label="france flag">
  {_t({ id: "language.fr" })}
</span>
```

**Pull request merged**